Framework/API/inc/MantidAPI/CitationConstructorHelpers.h

```diff
@@ -0,0 +1,239 @@
+#ifndef MANTID_API_CITATION_CITATIONCONSTRUCTORHELPERS_H_
+#define MANTID_API_CITATION_CITATIONCONSTRUCTORHELPERS_H_
+
+#include "DllConfig.h"
+
+#include <boost/optional.hpp>
+#include <vector>
+
+namespace Mantid {
+namespace API {
+
+struct MANTID_API_DLL BaseCitation {
```

Comment:
Why did you go with this approach rather then free functions that return `Citation`?

This approach has some rather annoying usability issues:
- it makes constructing a `Citation` more work (first having to create a helper object then create a `Citation` from that
- no automatic population of URL or DOI fields

Reply:
- You do not have to create a helper object first, it is 1 of 3 options. It said in the design document to add helper structs for easier construction of citations.
- All base citations should pass URL and or DOI into the respective Citation object.

These don't actually seem like problems.

I went with this approach as this way I don't need to rewrite 14 similar free functions, and instead I can utilise some basic polymorphism for easier implementation.

Reply:
> These don't actually seem like problems.

If I have to specify a URL in up to three places that is a problem.

Can you show a complete example of creating a citation for a paper (i.e. on that has all four entries)?

Reply:
```python
cite = mantid.api.citation(mantid.api.ArticleCitation(["author1"], "title", "journal", "year", "volume", "number", "pages", "month", "description", "doi", "url"))
url = cite.url()
```

```cpp
Mantid::API::Citation cite(Mantid::API::ArticleCitation({"author1"}, "title", "journal", "year", "volume", "number", "pages", "month", "description", "doi", "url")));
auto url  = cite.url();
```

Something like that will work.

Reply:
url in these cases will be "url".

Reply:
Ok, I see now.

However I feel this solution is overly complex and the polymorphism isn't actually gaining you anything (looking at the actual implementations of `toBibtex` and `toEndnote` backs this up).

Earlier we had discussed these conversion functions just being free functions that take the same parameters and return a `Citation`. What was limiting about this that made you implement the solution you present here?

Reply:
- There is very little cost to this
- As far as I am concerned 3 overloaded citation constructors is easier to understand than have 3 random functions.
- This follows object orientated design principles which Mantid does in the majority of its code

Reply:
- `CitationConstructorHelpers.h` and `CitationConstructorHelpers.cpp` contain more code than they would if you had used free functions in the way we had previously discussed
- OOP has a place, don't feel you have to use it just because Mantid already does in places
- OOP does not allow for much code reuse here and trying to shoehorn it in increases complexity

I say we leave this PR for now until we can discuss it in person next week.

Reply:
Can you please give me an example of why a free function implementation would be better?